### Spoken Language Processing
В этом задании предлагается обучить классификатор класса возраста по голосу (пример с тем, как это можно сделать для пола см. в семинаре)

Подумайте, как лучше предсказывать возраст (может быть разбить на группы?) и какой лосс использовать

P.S. не забудьте, что если то вы работает в Colab, то вы можете поменять среду выполнения на GPU/TPU!

Вопросы по заданию/материалам: @Nestyme

In [1]:
!pip3 install -q timit-utils==0.9.0
!pip3 install -q torchaudio
! wget https://ndownloader.figshare.com/files/10256148 
!unzip -q 10256148

--2022-08-12 10:03:02--  https://ndownloader.figshare.com/files/10256148
Resolving ndownloader.figshare.com (ndownloader.figshare.com)... 54.217.124.219, 52.16.102.173, 2a05:d018:1f4:d003:1c8b:1823:acce:812, ...
Connecting to ndownloader.figshare.com (ndownloader.figshare.com)|54.217.124.219|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://s3-eu-west-1.amazonaws.com/pfigshare-u-files/10256148/TIMIT.zip?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIYCQYOYV5JSSROOA/20220812/eu-west-1/s3/aws4_request&X-Amz-Date=20220812T100303Z&X-Amz-Expires=10&X-Amz-SignedHeaders=host&X-Amz-Signature=33b8f0ad95c627b411949b6b50f3a9c3b80b62190bb8835515f6cfc0412c36f2 [following]
--2022-08-12 10:03:03--  https://s3-eu-west-1.amazonaws.com/pfigshare-u-files/10256148/TIMIT.zip?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIYCQYOYV5JSSROOA/20220812/eu-west-1/s3/aws4_request&X-Amz-Date=20220812T100303Z&X-Amz-Expires=10&X-Amz-SignedHeaders=host&X-Amz-Signatur

In [2]:
import timit_utils as tu
import os
import librosa
import numpy as np
from tqdm import tqdm

import torch
import torch.nn as nn
from torch.optim import Adam
import torch.nn.functional as F

import matplotlib.pyplot as plt
from sklearn.metrics import accuracy_score

import IPython
_TIMIT_PATH = 'data/lisa/data/timit/raw/TIMIT'

In [3]:
device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
device

device(type='cpu')

## Задание 1
Загрузите данные для обучения. Для этого:
1. Скачайте датасет TIMIT (см семинар)
2. Соберите пары "голос"  — "класс возраста" также, как на семинаре собирались пары "голос"  — "пол". Аудиодорожки сконвертируйте в мелспектрограммы при помощи `torchaudio либо` `librosa`

P.S. вы можете использовать свою реализацию, а можете предложенную (см следующие ячейки)

In [18]:
import timit_utils as tu
import os
import librosa
import numpy as np
from tqdm import tqdm
import torch as t


class timit_dataloader:
    def __init__(self, data_path=_TIMIT_PATH, train_mode=True, age_mode=True):
        self.doc_file_path = os.path.join(data_path, 'DOC', 'SPKRINFO.TXT')
        self.corpus = tu.Corpus(data_path)
        with open(self.doc_file_path) as f:
            self.id_age_dict = dict(
                [(tmp.split(' ')[0], 86 - int(tmp.split('  ')[5].split('/')[-1].replace('??', '50'))) \
                 for tmp in f.readlines()[39:]])
        if train_mode:
            self.trainset = self.create_dataset('train', age_mode=age_mode)
            self.validset = self.create_dataset('valid', age_mode=age_mode)
        self.testset = self.create_dataset('test', age_mode=age_mode)

    def return_age(self, id):
        return self.id_age_dict[id]

    def return_data(self):
        return self.trainset, self.validset, self.testset

    def return_test(self):
        return self.testset

    def create_dataset(self, mode, age_mode=False):
        global people
        assert mode in ['train', 'valid', 'test']
        if mode == 'train':
            people = [self.corpus.train.person_by_index(i) for i in range(350)]
        if mode == 'valid':
            people = [self.corpus.train.person_by_index(i) for i in range(350, 400)]
        if mode == 'test':
            people = [self.corpus.test.person_by_index(i) for i in range(150)]
        spectrograms_and_targets = []
        for person in tqdm(people):
              try:
                  target = self.return_age(person.name)
                  for i in range(len(person.sentences)):
                      spectrograms_and_targets.append(
                          self.preprocess_sample(person.sentence_by_index(i).raw_audio, target, age_mode=True))
              except:
                  print(person.name, target)

        X, y = map(np.stack, zip(*spectrograms_and_targets))
        X = X.transpose([0, 2, 1])  # to [batch, time, channels]
        return X, y

    @staticmethod
    def spec_to_image(spec, eps=1e-6):
        mean = spec.mean()
        std = spec.std()
        spec_norm = (spec - mean) / (std + eps)
        spec_min, spec_max = spec_norm.min(), spec_norm.max()
        spec_scaled = 255 * (spec_norm - spec_min) / (spec_max - spec_min)
        spec_scaled = spec_scaled.astype(np.uint8)
        return spec_scaled

    @staticmethod
    def clasterize_by_age(age):
        if age < 18:
            return 0
        elif 18 <= age < 40:
            return 1
        elif 40 <= age < 60:
            return 2
        elif age >= 60:
            return 3

    def preprocess_sample(self, amplitudes, target, age_mode=False, sr=16000, max_length=150):
        spectrogram = librosa.feature.melspectrogram(amplitudes, sr=sr, n_mels=128, fmin=1, fmax=8192)[:, :max_length]
        spectrogram = np.pad(spectrogram, [[0, 0], [0, max(0, max_length - spectrogram.shape[1])]], mode='constant')
        target = self.clasterize_by_age(target)
        return self.spec_to_image(np.float32(spectrogram)), target

    def preprocess_sample_inference(self, amplitudes, sr=16000, max_length=150, device='cpu'):
        spectrogram = librosa.feature.melspectrogram(amplitudes, sr=sr, n_mels=128, fmin=1, fmax=8192)[:, :max_length]
        spectrogram = np.pad(spectrogram, [[0, 0], [0, max(0, max_length - spectrogram.shape[1])]], mode='constant')
        spectrogram = np.array([self.spec_to_image(np.float32(spectrogram))]).transpose([0, 2, 1])

        return t.tensor(spectrogram, dtype=t.float).to(device, non_blocking=True)


class dataloader:
    def __init__(self, spectrograms, targets):
        self.data = list(zip(spectrograms, targets))

    def next_batch(self, batch_size, device):
        indices = np.random.randint(len(self.data), size=batch_size)

        input = [self.data[i] for i in indices]

        source = [line[0] for line in input]
        target = [line[1] for line in input]
        batch, label = self.torch_batch(source, target, device)
        label = label.long()
        
        return batch, label

    @staticmethod
    def torch_batch(source, target, device):
        return tuple(
            [
                t.tensor(val, dtype=t.float).to(device, non_blocking=True)
                for val in [source, target]
            ]
        )

    @staticmethod
    def padd_sequences(lines, pad_token=0):
        lengths = [len(line) for line in lines]
        max_length = max(lengths)

        return np.array(
            [
                line + [pad_token] * (max_length - lengths[i])
                for i, line in enumerate(lines)
            ]
        )

Простая сверточная сеть, ее можно дотюнить или поменять по желанию

In [32]:
_timit_dataloader = timit_dataloader()
train, valid, test = _timit_dataloader.return_data()

  0%|          | 0/350 [00:00<?, ?it/s]

  0%|          | 0/50 [00:00<?, ?it/s]

  0%|          | 0/150 [00:00<?, ?it/s]

In [33]:
train_loader = dataloader(*train)
valid_loader = dataloader(*valid)
test_loader = dataloader(*test)

#Задание 2
1. Обучите свой классификатор категории возраста
2. Попробуйте улучшить результат. Можно попробовать усложнить сетку, подвигать границы категорий, поискать новые данные, что угодно, кроме учиться на тесте :)
3. Какой подход оказался самым эффективным? Как думаете, почему?
4. Как считаете, где можно было бы применить такой классификатор в качестве вспомогательной задачи?


In [46]:
import torch
import torch.nn as nn
import torch.nn.functional as F


N_CLASSES = 4

class Model(nn.Module):
    def __init__(self, window_sizes=(3, 4, 5, 6)):
        super(Model, self).__init__()

        self.convs = nn.ModuleList([
            nn.Conv2d(1, 256, [window_size, 128], padding=(window_size - 1, 0))
            for window_size in window_sizes
        ])

        self.fc = nn.Linear(256 * len(window_sizes), N_CLASSES)

    def forward(self, x):
        x = torch.unsqueeze(x, 1)  # [B, C, T, E] Add a channel dim.
        xs = []
        for conv in self.convs:
            x2 = F.relu(conv(x))  # [B, F, T, 1]
            x2 = torch.squeeze(x2, -1)  # [B, F, T]
            x2 = F.max_pool1d(x2, x2.size(2))  # [B, F, 1]
            xs.append(x2)
        x = torch.cat(xs, 2)  # [B, F, window]

        # FC
        x = x.view(x.size(0), -1)  # [B, F * window]
        logits = self.fc(x)  # [B, class]
        return logits

In [49]:
lr = 1e-4 # Предполагаемый learning rate. Он может быть больше или меньше :)

model = Model()
model.to(device)
optimizer = torch.optim.Adam(model.parameters(), lr)
criterion = torch.nn.CrossEntropyLoss()

In [50]:
from tqdm.notebook import tqdm

batch_size = 32
num_epochs = 10
# Train loop
for e in range(num_epochs):
    model.train()
    train_loss = 0
    for _ in tqdm(range(len(train_loader.data) // batch_size)):
        batch, label = train_loader.next_batch(batch_size, device)
        optimizer.zero_grad()
        output = model(batch)
        loss = criterion(output, label.to(device))
        train_loss += loss.item()
        loss.backward()
        optimizer.step()
        
        
    valid_loss = 0
    valid_acc = 0
    model.eval()
    with torch.no_grad():
        for _ in tqdm(range(len(valid_loader.data) // batch_size)):
            batch, label = valid_loader.next_batch(batch_size, device)
            optimizer.zero_grad()
            output = model(batch)
            loss = criterion(output, label.to(device))
            valid_loss += loss.item()
            valid_acc += (label.to(device) == output.argmax(axis=1)).sum().item()
            

    print(f"Train Loss: {train_loss / (len(train_loader.data))}, "
          f"Valid Loss: {valid_loss / (len(valid_loader.data))}, "
          f"Valid Acc: {valid_acc / (len(valid_loader.data))}")
    
        
# Testing
test_acc = 0
model.eval()
with torch.no_grad():
    for _ in tqdm(range(len(valid_loader.data) // batch_size)):
        batch, label = test_loader.next_batch(batch_size, device)
        output = model(batch.to(device))
        test_acc += (label.to(device) == output.argmax(axis=1)).sum().item()
        
print(f"\n\nTest Acc: {test_acc / len(test_loader.data)}")

  0%|          | 0/109 [00:00<?, ?it/s]

  0%|          | 0/15 [00:00<?, ?it/s]

Train Loss: 0.02146950314726148, Valid Loss: 0.030808187663555146, Valid Acc: 0.798


  0%|          | 0/109 [00:00<?, ?it/s]

  0%|          | 0/15 [00:00<?, ?it/s]

Train Loss: 0.012691473558545113, Valid Loss: 0.02159356790781021, Valid Acc: 0.782


  0%|          | 0/109 [00:00<?, ?it/s]

  0%|          | 0/15 [00:00<?, ?it/s]

Train Loss: 0.008753953207816396, Valid Loss: 0.021144441187381745, Valid Acc: 0.72


  0%|          | 0/109 [00:00<?, ?it/s]

  0%|          | 0/15 [00:00<?, ?it/s]

Train Loss: 0.0058694740281041175, Valid Loss: 0.031051153600215913, Valid Acc: 0.774


  0%|          | 0/109 [00:00<?, ?it/s]

  0%|          | 0/15 [00:00<?, ?it/s]

Train Loss: 0.004650888736758913, Valid Loss: 0.024639169335365297, Valid Acc: 0.786


  0%|          | 0/109 [00:00<?, ?it/s]

  0%|          | 0/15 [00:00<?, ?it/s]

Train Loss: 0.0032458376016999994, Valid Loss: 0.026035277009010316, Valid Acc: 0.784


  0%|          | 0/109 [00:00<?, ?it/s]

  0%|          | 0/15 [00:00<?, ?it/s]

Train Loss: 0.002346406975494964, Valid Loss: 0.019444404780864715, Valid Acc: 0.776


  0%|          | 0/109 [00:00<?, ?it/s]

  0%|          | 0/15 [00:00<?, ?it/s]

Train Loss: 0.0017868973830980914, Valid Loss: 0.021072945058345796, Valid Acc: 0.782


  0%|          | 0/109 [00:00<?, ?it/s]

  0%|          | 0/15 [00:00<?, ?it/s]

Train Loss: 0.0013725918337170566, Valid Loss: 0.0246086785197258, Valid Acc: 0.782


  0%|          | 0/109 [00:00<?, ?it/s]

  0%|          | 0/15 [00:00<?, ?it/s]

Train Loss: 0.0011080021070582525, Valid Loss: 0.019674346357584, Valid Acc: 0.784


  0%|          | 0/15 [00:00<?, ?it/s]



Test Acc: 0.262


In [53]:
pred = []
true = []
with torch.no_grad():
    for _ in tqdm(range(len(valid_loader.data) // batch_size)):
        batch, label = test_loader.next_batch(batch_size, device)
        output = model(batch.to(device))
        pred.append(output.argmax(axis=1).cpu().numpy())
        true.append(label.cpu().numpy())

  0%|          | 0/15 [00:00<?, ?it/s]

In [57]:
from sklearn.metrics import classification_report

print(classification_report(np.concatenate(true), np.concatenate(pred)))

              precision    recall  f1-score   support

           1       0.87      0.97      0.91       410
           2       0.35      0.13      0.19        62
           3       0.00      0.00      0.00         8

    accuracy                           0.84       480
   macro avg       0.40      0.36      0.37       480
weighted avg       0.79      0.84      0.80       480



/usr/local/lib/python3.7/dist-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.7/dist-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.7/dist-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


1. Самым эффективным подходом оказалось разделение на 3 группы людей с со границами следующими границами:
- 0-17
- 18-39 (точнее всего угадывается)
- 40-60
- 61+

2. Была усложнена модель, больше каналов (256) и еще дополнительный Conv слой с kernel 6.В результате accuracy на test вырос с 0.14 до 0.26